pkg/dataobj/internal/dataset/column_reader.go

```diff
@@ -0,0 +1,204 @@
+		}
+	}
+
+	for n < len(v) {
+		// Make sure our reader is initialized to the right page for the row we
+		// wish to read.
+		//
+		// If the next row is out of range of all pages, we return io.EOF here.
+		if cr.pageIndex == -1 || !cr.ranges[cr.pageIndex].Contains(uint64(cr.nextRow)) {
+			page, pageIndex, err := cr.nextPage()
+			if err != nil {
+				return n, err
+			}
+
+			switch cr.reader {
+			case nil:
+				cr.reader = newPageReader(page, cr.column.ColumnInfo().Type, cr.column.ColumnInfo().Compression)
+			default:
+				cr.reader.Reset(page, cr.column.ColumnInfo().Type, cr.column.ColumnInfo().Compression)
+			}
+
+			cr.pageIndex = pageIndex
+		}
+
+		// Ensure that our page reader is set to the correct row within the page.
+		//
+		// This seek is likely a no-op after the first iteration of the loop, but
+		// we call it each time anyway for safety.
+		pageRow := cr.nextRow - int64(cr.ranges[cr.pageIndex].Start)
+		if _, err := cr.reader.Seek(pageRow, io.SeekStart); err != nil {
+			// This should only return an error if pageRow < 0, shouldn't happen:
+			// cr.nextRow will always be >= cr.ranges[cr.pageIndex].Start.
+			return n, err
+		}
+
+		// Because len(v[n:]) shrinks with every iteration of our loop, one may be
+		// concerned that future iterations of the loop have less scratch space to
+		// work with, where scratch space is used to "skip" rows in the page.
+		//
+		// However, each call to [columnReader.Read] always reads a continguous set
+		// of rows, potentially across multiple page boundaries. This means that
+		// only the first call to cr.reader.Read will use the scratch space in v to
+		// skip rows, where the scratch space is the entirety of len(v).
+		count, err := cr.reader.Read(ctx, v[n:])
+		cr.nextRow += int64(count)
+		n += count
+
+		// We ignore io.EOF errors from the page; the next loop will detect and
+		// report EOF on the call to [columnReader.nextPage].
+		if err != nil && !errors.Is(err, io.EOF) {
+			return n, err
+		}
+	}
+
+	return n, nil
+}
+
+// nextPage returns the page where cr.nextRow is contained. If cr.nextRow is
+// beyond the boundaries of the column, nextPage returns nil, -1, io.EOF.
+func (cr *columnReader) nextPage() (Page, int, error) {
+	i := sort.Search(len(cr.ranges), func(i int) bool {
+		return cr.ranges[i].End >= uint64(cr.nextRow)
+	})
+	if i < len(cr.pages) && cr.ranges[i].Contains(uint64(cr.nextRow)) {
+		return cr.pages[i], i, nil
+	}
+	return nil, -1, io.EOF
+}
+
+// init initializes the set of pages and page row ranges.
+func (cr *columnReader) init(ctx context.Context) error {
+	var startRow uint64
+
+	for result := range cr.column.ListPages(ctx) {
+		page, err := result.Value()
+		if err != nil {
+			return err
+		}
+
+		endRow := startRow + uint64(page.PageInfo().RowCount) - 1
+
+		// TODO(rfratto): including page count in the column info metadata would
+		// allow us to set the capacity of cr.pages and cr.ranges more precisely.
+		cr.pages = append(cr.pages, page)
```

Comment:
Since we can reset and reuse this I think this isn't likely to be a big improvement. But makes sense. Let's create an issue.